kafka/src/main/java/nl/nn/adapterframework/extensions/kafka/KafkaInternalListener.java

```diff
@@ -0,0 +1,202 @@
+import nl.nn.adapterframework.receivers.RawMessageWrapper;
+import nl.nn.adapterframework.stream.Message;
+import nl.nn.adapterframework.stream.MessageContext;
+import nl.nn.adapterframework.util.StringUtil;
+
+import org.apache.kafka.common.serialization.ByteArrayDeserializer;
+import org.apache.kafka.common.serialization.StringDeserializer;
+
+/**
+ * This class should NOT be used outside of this kafka package.
+ * This class isn't public to prevent generation of javadoc/frankdoc.
+ * @param <T> Topic type.
+ * @param <M> Message type.
+ */
+class KafkaInternalListener<T,M> extends KafkaFacade implements IPullingListener<ConsumerRecord<T, M>> {
+	//setter is for testing purposes only.
+	private @Setter(AccessLevel.PACKAGE) Consumer<T, M> consumer;
+	private @Setter(AccessLevel.PACKAGE) Function<Properties, Consumer<T,M>> consumerGenerator = KafkaConsumer::new;
+	private Iterator<ConsumerRecord<T, M>> waiting;
+	private @Setter String groupId;
+	private @Setter boolean fromBeginning;
+	private @Setter int patternRecheckInterval;
+	private @Setter String topics;
+	private @Getter(AccessLevel.PACKAGE) List<Pattern> topicPatterns;
+	private final Duration pollDuration = Duration.ofMillis(1);
+	private final Map<TopicPartition, OffsetAndMetadata> offsetAndMetadataMap = new HashMap<>();
+	private final BiFunction<M, MessageContext, Message> converter;
+	private final Lock lock = new ReentrantLock();
+	private final String keyDeserializer;
+	private final String valueDeserializer;
+
+	public KafkaInternalListener(Properties properties, KafkaType keyType, KafkaType messageType) {
+		super();
+		this.properties = properties;
+		this.converter = messageConverterFactory(messageType);
+		this.keyDeserializer = getDeserializer(keyType);
+		this.valueDeserializer = getDeserializer(messageType);
+	}
+
+	private static String getDeserializer(KafkaType type) {
+		if(type==KafkaType.STRING) return StringDeserializer.class.getName();
+		if(type==KafkaType.BYTEARRAY) return ByteArrayDeserializer.class.getName();
+		throw new IllegalArgumentException("Unknown KafkaType ["+type+"]");
+	}
+
+	public static <M> BiFunction<M, MessageContext, Message> messageConverterFactory(KafkaType kafkaType) {
+		if(kafkaType == KafkaType.STRING) return (M message, MessageContext messageContext) -> new Message((String) message, messageContext);
+		if(kafkaType == KafkaType.BYTEARRAY) return (M message, MessageContext messageContext) -> new Message((byte[]) message, messageContext);
+		throw new IllegalArgumentException("Unknown KafkaType ["+kafkaType+"]");
+	}
+	@Override
+	public void configure() throws ConfigurationException {
+		super.configure();
+		if (StringUtils.isEmpty(groupId)) throw new ConfigurationException("groupId must be specified");
+		if (StringUtils.isEmpty(topics)) throw new ConfigurationException("topics must be specified");
+		if(patternRecheckInterval < 10) throw new ConfigurationException("patternRecheckInterval should be at least 10");
```

Comment:
"Should be at least 10", but the default right now appears to be 0.
What is a sensible default, and can we set it to that default when we initialise the field?

Reply:
It's a bit confusing, but the defaults are set in the `KafkaListener`. Here, the default is set to `5000`.

This field is copied over to the `KafkaInternalListener` in the `configure` method.

Reply:
OK -- that is very confusing and I do not see why we need this extra layer of classes.